packages/@lwc/integration-karma/test/component/dynamic-component/index.spec.js

```diff
@@ -0,0 +1,267 @@
+import { createElement } from 'lwc';
+import Attributes from 'x/attributes';
+import Basic from 'x/basic';
+import Slotter from 'x/slotter';
+import List from 'x/list';
+import Nested from 'x/nested';
+import ScopedSlotParent from 'x/scopedSlotParent';
+import ForwardedScopedSlotParent from 'x/forwardedScopedSlotParent';
+import Conditional from 'x/conditional';
+
+describe('dynamic components', () => {
+    it('basic dynamic component', () => {
+        const elm = createElement('x-basic', { is: Basic });
+        document.body.appendChild(elm);
+
+        return Promise.resolve()
+            .then(() => {
+                // Constructor not set
+                const children = elm.shadowRoot.children;
+                expect(children.length).toBe(0);
+                elm.loadFoo();
+            })
+            .then(() => {
+                // Set constructor to foo
+                const children = elm.shadowRoot.children;
+                expect(children.length).toBe(1);
+                expect(children[0].tagName.toLowerCase()).toBe('x-foo');
+                elm.loadBar();
+            })
+            .then(() => {
+                // Change constructor to bar
+                const children = elm.shadowRoot.children;
+                expect(children.length).toBe(1);
+                expect(children[0].tagName.toLowerCase()).toBe('x-bar');
+                elm.clearCtor();
+            })
+            .then(() => {
+                // Change constructor to null
+                const children = elm.shadowRoot.children;
+                expect(children.length).toBe(0);
+            });
+    });
+
+    it('attributes assigned to dynamic components', () => {
+        const elm = createElement('x-attributes', { is: Attributes });
+        document.body.appendChild(elm);
+
+        return Promise.resolve().then(() => {
+            const foo = elm.shadowRoot.querySelector('x-foo');
+            const bar = elm.shadowRoot.querySelector('x-bar');
+
+            expect(foo).not.toBeNull();
+            expect(bar).not.toBeNull();
+
+            spyOn(console, 'log');
+
+            // declaratively assigned
+            expect(foo.className).toContain('slds-snazzy');
+            foo.click();
+            // eslint-disable-next-line no-console
+            expect(console.log).toHaveBeenCalledWith('template click called');
```

Comment:
If the goal is to test that events propagate correctly up from a dynamic component to a parent, I would recommend:

1. Using a custom event with `dispatchEvent` with `{ bubbles: true, composed: true }` rather than the `click` event, and
2. Adding an `@api` to the `x-attributes` component that returns the evidence that the event was received

Otherwise, we are mucking with spies on `console.log` (which may interact with other stuff that relies on spying on console), and synthetic `click`s are always a bit dodgy because the browser treats them a bit differently from user-initiated clicks.

Reply:
Instead of the `@api`, you could also have some kind of queue like `window.mySpecialTestQueue = []`. We've done this kind of thing for lifecycle callback timing.

Reply:
You know what, I just realized we are already doing `elm.click()` in plenty of places in our Karma tests, so forget what I said about that. But I do think mocking the `console.log` seems unnecessary.

Reply:
Great feedback!  After reading through your response, I think you're right testing off of `console.log` does seem unnecessary.  

I took your suggestion and verify the event using `event.target` instead, here's what the [new tests](https://github.com/salesforce/lwc/pull/3355/commits/1658d36d1a001bfce3a4a441323e9d4caf473f48#diff-de91f87325f57e40b7726e4008cc192262ccc8203409683983917c29ac490d4fR58-R66) look like.